### PR TITLE
[ticket/16913] Add Search Index Progress Bar with Stats

### PR DESCRIPTION
**Pull request by Dark1z**
Checklist:

- [x] Correct branch: master for new features; 3.3.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/master/coding-guidelines.html) and [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

ScreenShot:
.
1st run (On Submit) : 

.
2nd & consecutive run (On Refresh) : 

.

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-16913

### CODE REVIEW

**Comment by rxu** [comment]
What about already existing progress bar?
https://github.com/phpbb/phpbb/blob/3.3.x/phpBB/adm/style/acp_search.html#L157

**Comment by Dark1z** [comment]
> What about already existing progress bar? https://github.com/phpbb/phpbb/blob/3.3.x/phpBB/adm/style/acp_search.html#L157

ScreenShots :
1.

2.

.
Best regards 👍

**Comment by rxu** [comment]
And? :) That doesn't do anything with the current progress bar so phpBB will get 2 for some reason.

File: phpBB/language/en/acp/search.php [hunk 1]
```diff
@@ -125,6 +125,7 @@
 	),
 	'SEARCH_INDEX_CREATED'					=> 'Successfully indexed all posts in the board database.',
 	'SEARCH_INDEX_REMOVED'					=> 'Successfully deleted the search index for this backend.',
+	'SEARCH_INDEX_PROGRESS'					=> '<br><progress value="%1$d" max="%3$d"></progress><br> %4$.2f %% <br>Done: %1$d | Pending: %2$d | Total: %3$d',
```
Comment (rxu) [inline phpBB/language/en/acp/search.php:128]: It'd be better to have all HTML markup in template rather than on PHP.

**Comment by Dark1z** [comment]
> And? :) That doesn't do anything with the current progress bar so phpBB will get 2 for some reason.

the one which you are mentioning can be called "Loading Bar",
which loads this template file:
https://github.com/phpbb/phpbb/blob/3.3.x/phpBB/adm/style/progress_bar.html#L30
which in turn loads this GIF:
https://github.com/phpbb/phpbb/blob/3.3.x/phpBB/adm/images/progress_bar.gif

File: phpBB/language/en/acp/search.php (hunk 1 above)
Comment (Dark1z) [reply phpBB/language/en/acp/search.php:128]: can not do that in `trigger_error(...)` function.

Comment (Dark1z) [reply phpBB/language/en/acp/search.php:128]: do you have a way to do that ???

Comment (rxu) [reply phpBB/language/en/acp/search.php:128]: Well, in theory that cab be done via adding relevant code to `\adm\style\message_body.html` and assigning related template vars before `trigger_error()` call.

Comment (Dark1z) [reply phpBB/language/en/acp/search.php:128]: since it's a small thing,
and also the `%1$d` & `%3$d` are repetitive,
hence I added it in the lang file.

**Comment by DavidIQ** [comment]
We are moving away from adding/including html in language variables so adding more of it is going to be a non-starter for this PR.

**Comment by Dark1z** [comment]
> We are moving away from adding/including html in language variables so adding more of it is going to be a non-starter for this PR.

Then is it ok to move it to php file ???

**Comment by rubencm** [comment]
I like the progress bar, but I would do it when the data you already have: `max_post_id` and `post_counter` to avoid add extra queries

**Comment by DavidIQ** [comment]
> > We are moving away from adding/including html in language variables so adding more of it is going to be a non-starter for this PR.
> 
> Then is it ok to move it to php file ???

In general HTML should go in template files.

**Comment by Dark1z** [comment]
> I like the progress bar, but I would do it when the data you already have: `max_post_id` and `post_counter` to avoid add extra queries

I did not get what you mean?
how am I going to get the `$done_count`, `$remain_count` & `$total_count` from `max_post_id` and `post_counter` ?

**Comment by Dark1z** [comment]
> In general HTML should go in template files.

Yes! But that's not possible via `trigger_error(...)`,
so I have to inject similar to this:
https://github.com/phpbb/phpbb/pull/6328/files#diff-2f31352ab73095253f68c8cd43b1065cd218084bbadea4deacd66deec5d3e0f2R556-R563

**Comment by 3D-I** [comment]
I am not sure that this adds to or improves the current statistics, I find it at least just a matter of aesthetics not entirely necessary. Rather I would be in favour of having a slide that updates ajaxed and certainly not on page refresh (popup). Unless I am missing something?

**Comment by Dark1z** [comment]
> I am not sure that this adds to or improves the current statistics, I find it at least just a matter of aesthetics not entirely necessary. Rather I would be in favour of having a bar that updates ajaxed and certainly not on page refresh (popup). Unless I am missing something? If the goal is to increase the size of the progress bar by 2x, then IMHO it is very trivial. Not to mention the HTML inclusion, which is somehow a "no-no".

Have you never done the search index on 100K post or more board ?

It's default behavior is to page refresh, I am not changing that.
I only add a progress bar and some stats.

This isn't a "trivial" on large boards.

If you have better solution to "HTML inclusion", then kindly make a PR for the same.

Best regards 👍

**Comment by 3D-I** [comment]
> Have you never done the search index on board with 100K or more post ?

In the last 15 years I think at least a couple (dozens) of times or more. I don't see the connection anyway. 

If a PR in my opinion is almost superfluous, I say so expressing my opinion, then I understand that it can be annoying but I don't think I should limit my thoughts.

**Comment by Dark1z** [comment]
> In the last 15 years I think at least a couple (dozens) of times or more. I don't see the connection anyway.
> 
> If a PR in my opinion is almost superfluous, I say so expressing my opinion, then I understand that it can be annoying but I don't think I should limit my thoughts.


Also, I am not touching that Pop-Up loading progress bar,
IMHO, I do not know the use of that, or why is it used for, so did not touch it.

Best regards 👍

**Comment by 3D-I** [comment]
> IMHO, I do not know the use of that, or why is it used for, so did not touch it.

Enough said 😃

**Comment by Dark1z** [comment]
> Enough said 😃

I think you are referring to this part of code : 
https://github.com/phpbb/phpbb/blob/ceca69d121fbb5b5952a9f65eb567e54030aaca7/phpBB/includes/acp/acp_search.php#L536-L554
Which generates the Pop-Up loading progress bar,
for which (again) I have not touched.

Best regards 👍

**Comment by 3D-I** [comment]
Again,  I find this PR not necessary. Period. Thanks.

**Comment by Dark1z** [comment]
> Again, I find this PR not necessary. Period. Thanks.

But I find this PR necessary. Period. Thanks.

**Comment by Dark1z** [comment]
!set wip

**Comment by GanstaZ** [comment]
Don't get me wrong.. I'm not against progress bar update, but adding more of those trigger errors into modules (controllers in the future) is not good as those will be removed.

**Comment by Dark1z** [comment]
> Don't get me wrong.. I'm not against progress bar update, but adding more of those trigger errors into modules (controllers in the future) is not good as those will be removed.

Is there any alternatives ?

Best regards 👍

**Comment by GanstaZ** [comment]
Hard to say.. master branch compared to 3.3.x differs a lot.. In master version i removed all triggers from acp search with a solution i use in other files, but it wont work on 3.3.x.

**Comment by Dark1z** [comment]
!unset WIP

**Comment by Dark1z** [comment]
https://github.com/phpbb/phpbb/pull/6328#issuecomment-1036459835
> Hard to say.. master branch compared to 3.3.x differs a lot.. In master version i removed all triggers from acp search with a solution i use in other files, but it wont work on 3.3.x.

@GanstaZ ,

can I get references to those commits or PR of master version,
I currently need them for master version of this PR.

Best regards 👍

**Comment by GanstaZ** [comment]
> [#6328 (comment)](https://github.com/phpbb/phpbb/pull/6328#issuecomment-1036459835)
> 
> can I get references to those commits or PR of master version, I currently need them for master version of this PR.
> 
> Best regards 👍

My PR progress is frozen because of health issues.. Best thing you can do, is to discuss your changes with phpBB team.

**Pull request closed without merging**